src/encryption/mod.rs

```diff
@@ -9,62 +12,79 @@ pub struct AES {
 
 }
 
-impl AES {
-  pub fn encrypt_128_impl(key: &[u8], iv: &[u8], message: &[u8]) -> Result<Vec<u8>> {
-    let cipher = Cbc::<Aes128, Pkcs7>::new_from_slices(key, iv)?;
-
-    Ok(cipher.encrypt_vec(message))
-  } 
-
-  pub fn decrypt_128_impl(key: &[u8], iv: &[u8], message: &[u8]) -> Result<Vec<u8>> {
-    let cipher = Cbc::<Aes128, Pkcs7>::new_from_slices(key, iv)?;
-
-    Ok(cipher.decrypt_vec(&message)?)
-  }
-
-  pub fn encrypt_256_impl(key: &[u8], iv: &[u8], message: &[u8]) -> Result<Vec<u8>> {
-    let cipher = Cbc::<Aes256, Pkcs7>::new_from_slices(key, iv)?;
+#[wasm_bindgen]
+#[derive(Debug, Clone, Copy)]
+pub enum AESAlgorithms {
+   AES128_CBC,
+   AES256_CBC,
+   AES128_CTR,
+   AES256_CTR
+}
 
-    Ok(cipher.encrypt_vec(&message))
+impl AES {
+  pub fn encrypt_impl(key: &[u8], iv: &[u8], message: &[u8], algo: AESAlgorithms) -> Result<Vec<u8>> {
+    let result = match algo {
+        AESAlgorithms::AES128_CBC => Cbc::<Aes128, Pkcs7>::new_from_slices(key, iv)?.encrypt_vec(&message),
+        AESAlgorithms::AES256_CBC => Cbc::<Aes256, Pkcs7>::new_from_slices(key, iv)?.encrypt_vec(&message),
+        AESAlgorithms::AES128_CTR => {
+            //let mut data = *message.clone();
+            let mut data = [1, 2, 3, 4, 5, 6, 7];
+            let mut cipher = Aes128Ctr::new(key.into(), iv.into());
+            cipher.apply_keystream(&mut data);
+            return Ok(data.to_vec());
+        }
+        AESAlgorithms::AES256_CTR => {
+            //let mut data = *message.clone();
+            let mut data = [1, 2, 3, 4, 5, 6, 7];
+            let mut cipher = Aes256Ctr::new(key.into(), iv.into());
+            cipher.apply_keystream(&mut data);
+            return Ok(data.to_vec());
+        }
+    };
+
+    Ok(result)
   }
 
-  pub fn decrypt_256_impl(key: &[u8], iv: &[u8], message: &[u8]) -> Result<Vec<u8>> {
-    let cipher = Cbc::<Aes256, Pkcs7>::new_from_slices(key, iv)?;
-
-    Ok(cipher.decrypt_vec(&message)?)
+  pub fn decrypt_impl(key: &[u8], iv: &[u8], message: &[u8], algo: AESAlgorithms) -> Result<Vec<u8>> {
+    let result = match algo {
+        AESAlgorithms::AES128_CBC => Cbc::<Aes128, Pkcs7>::new_from_slices(key, iv)?.decrypt_vec(&message),
+        AESAlgorithms::AES256_CBC => Cbc::<Aes256, Pkcs7>::new_from_slices(key, iv)?.decrypt_vec(&message),
+        AESAlgorithms::AES128_CTR => {
+            //let mut data = *message.clone();
+            let mut data = [168, 3, 171, 3, 110, 163, 29];
```

Comment:
@Firaenix currently hardcoded the data here on ctr mode - getting compile issues when trying to pass the message itself as mutable

Reply:
sorted. merge my `aes` branch into this PR